lib/c-api/src/wasm_c_api/externals/function.rs

```diff
@@ -122,8 +122,12 @@ pub unsafe extern "C" fn wasm_func_new_with_env(
 
     impl Drop for WrapperEnv {
         fn drop(&mut self) {
-            if let Some(env_finalizer) = *self.env_finalizer {
-                unsafe { (env_finalizer)(self.env as _) }
+            if Arc::strong_count(&self.env_finalizer) == 1 {
+                if let Ok(mut guard) = self.env_finalizer.lock() {
```

Comment:
Should we lock before the strong count to prevent double drops?

Reply:
There's not gonna be a double drop, because the callback is behind a mutex. I think what you mean to say is that strong_count might return 2 and before that is compared to 1, another instance has dropped. So the only failing scenario is the callback not running at all. I will switch the lock to be first.

Reply:
Yup, that's what I was referring. Thanks!